build_test.go

```diff
@@ -0,0 +1,386 @@
+
+	context("when the checksum for composer.lock matches a previous layer's checksum", func() {
+		it.Before(func() {
+			buildpackPlan.Entries[0].Metadata["launch"] = true
+			buildpackPlan.Entries[0].Metadata["build"] = true
+			calculator.SumCall.Returns.String = "sha-from-composer-lock"
+
+			err := ioutil.WriteFile(filepath.Join(layersDir, fmt.Sprintf("%s.toml", composer.ComposerPackagesLayerName)),
+				[]byte(`[metadata]
+composer-lock-sha = "sha-from-composer-lock"
+`), os.ModePerm)
+			Expect(err).NotTo(HaveOccurred())
+
+		})
+
+		it("reuses the cached version of the SDK dependency", func() {
+			result, err := build(packit.BuildContext{
+				WorkingDir: workingDir,
+				Layers:     packit.Layers{Path: layersDir},
+				Plan:       buildpackPlan,
+			})
+			Expect(err).NotTo(HaveOccurred())
+
+			Expect(buffer).NotTo(ContainSubstring("Running 'composer install'"))
+
+			Expect(calculator.SumCall.Receives.Paths).To(Equal([]string{filepath.Join(workingDir, "composer.lock")}))
+
+			Expect(result).To(Equal(packit.BuildResult{
+				Layers: []packit.Layer{
+					{
+						Name:             composer.ComposerPackagesLayerName,
+						Path:             filepath.Join(layersDir, composer.ComposerPackagesLayerName),
+						SharedEnv:        packit.Environment{},
+						BuildEnv:         packit.Environment{},
+						LaunchEnv:        packit.Environment{},
+						ProcessLaunchEnv: map[string]packit.Environment{},
+						Build:            true,
+						Launch:           true,
+						Cache:            true,
+						Metadata: map[string]interface{}{
+							"composer-lock-sha": "sha-from-composer-lock",
+						},
+					},
+				},
+			}))
+		})
+	})
+
+	context("invokes 'composer check-platform-reqs'", func() {
+		it("generates 'php.ini.d/composer-extensions.ini'", func() {
+			_, err := build(packit.BuildContext{
+				WorkingDir: workingDir,
+				Layers:     packit.Layers{Path: layersDir},
+				Plan:       buildpackPlan,
+			})
+			Expect(err).NotTo(HaveOccurred())
+
+			Expect(composerCheckPlatformReqsExecExecution.Args[0]).To(Equal("check-platform-reqs"))
+			Expect(composerCheckPlatformReqsExecExecution.Dir).To(Equal(workingDir))
+			Expect(len(composerCheckPlatformReqsExecExecution.Env)).To(Equal(len(os.Environ()) + 4))
+
+			Expect(composerCheckPlatformReqsExecExecution.Env).To(ContainElements(
+				"COMPOSER_NO_INTERACTION=1",
+				fmt.Sprintf("COMPOSER_HOME=%s", filepath.Join(layersDir, composer.ComposerPackagesLayerName, ".composer")),
+				fmt.Sprintf("PHPRC=%s", filepath.Join(layersDir, "composer-php-ini", "composer-php.ini")),
+				fmt.Sprintf("PATH=fake-path-from-tests")))
+
+			Expect(filepath.Join(workingDir, "php.ini.d", "composer-extensions.ini")).To(BeARegularFile())
+
+			contents, err := os.ReadFile(filepath.Join(workingDir, "php.ini.d", "composer-extensions.ini"))
+			Expect(err).NotTo(HaveOccurred())
+
+			Expect(string(contents)).To(Equal(`extension = hello.so
+extension = bar.so
+`))
+		})
+	})
+
+	context("failure cases", func() {
```

Comment:
We're missing quite a few possible failure case tests. Is there a reason you've omitted them>?

Reply:
Some of the failure cases aren't really reachable or at least I can't find a way to make them reachable in the tests, or possibly I didn't feel like the lifecycle would give us that scenario.

For example, if `layersDir` is unwriteable, `Build` will fail pretty quickly, but probably 10 or 12 failure cases are hidden behind that first error. That being said, I'll try to increase coverage.

Reply:
Nitpick, for unreachable cases that are left, can you add an `// untested` comment around the untested line? I've seen this in a few buildpacks and I find it helpful.